### ISSUE TITLE
Unused local function is not reported in case of recursion (CS8321)

### ISSUE THREAD

**Issue by JosefPihrt** [dotnet/roslyn#29391]
**Version Used**: VS 15.8

**Steps to Reproduce**:

```csharp
namespace N
{
    class C
    {
        void M()
        {
            void F() // CS8321
            {
            }
        }

        void M2()
        {
            void F() // No warning here
            {
                F();
            }
        }
    }
}
```

**Comment by jaredpar** [comment]
I will let @agocke confirm but I believe this behavior is By Design. Consider a similar example with locals that also does not warn:

``` csharp
using System;
public class C {
    public void M() {
        int x = x = 42;
    }
}
```

**Comment by agocke** [comment]
Yup, this is by design.
